File: sdk-rust/crates/dex-sdk/tests/integ/rpc_workflow.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::sync::LazyLock;

use dex_sdk::{
    Attribute, AttributeIndex, AttributeMap, Channel, Context, Flow, HandlerError, HandlerResult,
    PersistenceSchema, Rpc, RpcList, RpcResult, Step, StepDecision, StepList, Wait,
};

pub(crate) static DATA: LazyLock<Attribute<String>> = LazyLock::new(|| Attribute::new("rpc-data"));
pub(crate) static KEYWORD: LazyLock<Attribute<String>> =
    LazyLock::new(|| Attribute::new("CustomKeywordField").indexed(AttributeIndex::keyword()));
pub(crate) static INTEGER: LazyLock<Attribute<i32>> =
    LazyLock::new(|| Attribute::new("CustomIntField").indexed(AttributeIndex::int()));
pub(crate) static INTERNAL: LazyLock<Channel<()>> = LazyLock::new(|| Channel::new("rpc-internal"));

pub(crate) struct RpcWorkflow {
    map: AttributeMap<String>,
    first: FirstStep,
    output: OutputStep,
}

impl RpcWorkflow {
    pub(crate) const RPC_OUTPUT: i64 = 100;
    pub(crate) const HARDCODED_VALUE: &str = "random-string";
    pub(crate) const PUBLISH_WITHOUT_ATTRIBUTE_ACCESS: Rpc<(), ()> =
        Rpc::new("publish_without_attribute_access");
    pub(crate) const FUNCTION_ONE: Rpc<String, i64> = Rpc::new("function_one");
    pub(crate) const FUNCTION_ZERO: Rpc<(), i64> = Rpc::new("function_zero");
    pub(crate) const PROCEDURE_ONE: Rpc<String, ()> = Rpc::new("procedure_one");
    pub(crate) const PROCEDURE_ZERO: Rpc<(), ()> = Rpc::new("procedure_zero");
    pub(crate) const READ_ONLY: Rpc<String, i64> = Rpc::new("read_only");
    pub(crate) const SET_DATA: Rpc<Option<String>, ()> = Rpc::new("set_data");
    pub(crate) const GET_DATA: Rpc<(), Option<String>> = Rpc::new("get_data");
    pub(crate) const SET_KEYWORD: Rpc<Option<String>, ()> = Rpc::new("set_keyword");
    pub(crate) const GET_KEYWORD: Rpc<(), Option<String>> = Rpc::new("get_keyword");

    pub(crate) fn new() -> Self {
        Self {
            first: FirstStep,
            output: OutputStep,
            map: AttributeMap::new("rpc-map"),
        }
    }

    fn publish_without_attribute_access(&self, context: &mut Context) -> HandlerResult<()> {
        Self::require_context(context)?;
        INTERNAL.publish(context, ())
    }

    fn function_one(&self, context: &mut Context, input: String) -> HandlerResult<RpcResult<i64>> {
        Self::require_context(context)?;
        DATA.delete(context)?;
        DATA.set(context, input.clone())?;
        KEYWORD.set(context, input)?;
        INTEGER.set(context, Self::RPC_OUTPUT as i32)?;
        INTERNAL.publish(context, ())?;
        Ok(RpcResult::new(Self::RPC_OUTPUT))
    }

    fn function_zero(&self, context: &mut Context) -> HandlerResult<RpcResult<i64>> {
        Self::require_context(context)?;
        DATA.set(context, Self::HARDCODED_VALUE.into())?;
        KEYWORD.set(context, Self::HARDCODED_VALUE.into())?;
        INTEGER.set(context, Self::RPC_OUTPUT as i32)?;
        INTERNAL.publish(context, ())?;
        Ok(RpcResult::new(Self::RPC_OUTPUT))
    }

    fn procedure_one(&self, context: &mut Context, input: String) -> HandlerResult<()> {
        Self::require_context(context)?;
        DATA.set(context, input.clone())?;
        KEYWORD.set(context, input)?;
        INTEGER.set(context, Self::RPC_OUTPUT as i32)?;
        INTERNAL.publish(context, ())
    }

    fn procedure_zero(&self, context: &mut Context) -> HandlerResult<()> {
        Self::require_context(context)?;
        DATA.set(context, Self::HARDCODED_VALUE.into())?;
        KEYWORD.set(context, Self::HARDCODED_VALUE.into())?;
        INTEGER.set(context, Self::RPC_OUTPUT as i32)?;
        INTERNAL.publish(context, ())
    }

    fn read_only(&self, context: &mut Context, _input: String) -> HandlerResult<RpcResult<i64>> {
        Self::require_context(context)?;
        Ok(RpcResult::new(Self::RPC_OUTPUT))
    }

    fn set_data(&self, context: &mut Context, input: Option<String>) -> HandlerResult<()> {
        Self::set_optional(context, &DATA, input)
    }

    fn get_data(&self, context: &mut Context) -> HandlerResult<RpcResult<Option<String>>> {
        Ok(RpcResult::new(DATA.get(context)?))
    }

    fn set_keyword(&self, context: &mut Context, input: Option<String>) -> HandlerResult<()> {
        Self::set_optional(context, &KEYWORD, input)
    }

    fn get_keyword(&self, context: &mut Context) -> HandlerResult<RpcResult<Option<String>>> {
        Ok(RpcResult::new(KEYWORD.get(context)?))
    }

    fn set_optional(
        context: &mut Context,
        attribute: &Attribute<String>,
        value: Option<String>,
    ) -> HandlerResult<()> {
        match value {
            Some(value) => attribute.set(context, value),
            None => attribute.delete(context),
        }
    }

    fn require_context(context: &Context) -> HandlerResult<()> {
        if context.flow_id().is_empty() || context.run_id().is_empty() {
            return Err(HandlerError::new("RpcFailure", "invalid RPC context"));
        }
        Ok(())
    }
}

impl Flow for RpcWorkflow {
    type StartInput = i32;

    fn steps(&self) -> StepList<'_, Self::StartInput> {
        StepList::start(&self.first).and(&self.output)
    }

    fn persistence(&self) -> PersistenceSchema {
        PersistenceSchema::new()
            .attribute(&DATA)
            .attribute(&KEYWORD)
            .attribute(&INTEGER)
            .attribute_map(&self.map)
            .channel(&INTERNAL)
    }

    fn rpcs(&self) -> RpcList<Self> {
        RpcList::new()
            .procedure_without_input(
                Self::PUBLISH_WITHOUT_ATTRIBUTE_ACCESS,
                Self::publish_without_attribute_access,
            )
            .function(Self::FUNCTION_ONE, Self::function_one)
            .function_without_input(Self::FUNCTION_ZERO, Self::function_zero)
            .procedure(Self::PROCEDURE_ONE, Self::procedure_one)
            .procedure_without_input(Self::PROCEDURE_ZERO, Self::procedure_zero)
            .function(Self::READ_ONLY, Self::read_only)
            .procedure(Self::SET_DATA, Self::set_data)
            .function_without_input(Self::GET_DATA, Self::get_data)
            .procedure(Self::SET_KEYWORD, Self::set_keyword)
            .function_without_input(Self::GET_KEYWORD, Self::get_keyword)
    }
}

struct FirstStep;

impl Step for FirstStep {
    type Input = i32;

    fn wait_for(&self, _context: &mut Context, _input: i32) -> HandlerResult<Wait> {
        Ok(Wait::until(INTERNAL.for_one()))
    }

    fn execute(&self, _context: &mut Context, _input: i32) -> HandlerResult<StepDecision> {
        Ok(StepDecision::go_to(&OutputStep, 0))
    }
}

struct OutputStep;

impl Step for OutputStep {
    type Input = i32;

    fn execute(&self, _context: &mut Context, _input: i32) -> HandlerResult<StepDecision> {
        Ok(StepDecision::graceful_complete(2_i32))
    }
}

File: sdk-java/src/test/java/io/superdurable/dex/integ/RpcWorkflow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.AttributeIndex;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.Wait;

class RpcWorkflow implements Flow<Integer> {
    static final long RPC_OUTPUT = 100L;
    static final String HARDCODED_VALUE = "random-string";
    final Channel<Void> internal = Channel.define("rpc-internal", Void.class);
    final Attribute<String> data = Attribute.define("rpc-data", String.class);
    final Attribute<String> keyword = Attribute.define(
            "CustomKeywordField",
            String.class,
            new AttributeIndex(AttributeIndex.Type.KEYWORD));
    final Attribute<Integer> integer = Attribute.define(
            "CustomIntField",
            Integer.class,
            new AttributeIndex(AttributeIndex.Type.INT));
    private final RpcFirstStep first = new RpcFirstStep();
    private final RpcOutputStep output = new RpcOutputStep();

    @Override
    public StepList<Integer> getSteps() {
        return StepList.startStep(first).otherSteps(output);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(data, keyword, integer, internal);
    }

    @RPC
    public void publishWithoutAttributeAccess(final Context context) {
        requireContext(context);
        internal.publish(context, null);
    }

    @RPC
    public RPCResult<Long> functionOne(final Context context, final String input) {
        requireContext(context);
        data.set(context, null);
        data.set(context, input);
        keyword.set(context, input);
        integer.set(context, Math.toIntExact(RPC_OUTPUT));
        internal.publish(context, null);
        return RPCResult.of(RPC_OUTPUT);
    }

    @RPC
    public RPCResult<Long> functionZero(final Context context) {
        requireContext(context);
        data.set(context, HARDCODED_VALUE);
        keyword.set(context, HARDCODED_VALUE);
        integer.set(context, Math.toIntExact(RPC_OUTPUT));
        internal.publish(context, null);
        return RPCResult.of(RPC_OUTPUT);
    }

    @RPC
    public void procedureOne(final Context context, final String input) {
        requireContext(context);
        data.set(context, input);
        keyword.set(context, input);
        integer.set(context, Math.toIntExact(RPC_OUTPUT));
        internal.publish(context, null);
    }

    @RPC
    public void procedureZero(final Context context) {
        requireContext(context);
        data.set(context, HARDCODED_VALUE);
        keyword.set(context, HARDCODED_VALUE);
        integer.set(context, Math.toIntExact(RPC_OUTPUT));
        internal.publish(context, null);
    }

    @RPC
    public RPCResult<Long> readOnly(final Context context, final String input) {
        requireContext(context);
        return RPCResult.of(RPC_OUTPUT);
    }

    @RPC
    public void setData(final Context context, final String input) {
        requireContext(context);
        data.set(context, input);
    }

    @RPC
    public RPCResult<String> getData(final Context context) {
        requireContext(context);
        return RPCResult.of(data.get(context));
    }

    @RPC
    public void setKeyword(final Context context, final String input) {
        requireContext(context);
        keyword.set(context, input);
    }

    @RPC
    public RPCResult<String> getKeyword(final Context context) {
        requireContext(context);
        return RPCResult.of(keyword.get(context));
    }

    private static void requireContext(final Context context) {
        if (context.getFlowId().isEmpty() || context.getRunId().isEmpty()) {
            throw new IllegalStateException("invalid RPC context");
        }
    }

    final class RpcFirstStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public Wait waitFor(final Context context, final Integer input) {
            return Wait.until(internal.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            return StepDecision.goTo(RpcOutputStep.class, 0);
        }
    }

    static final class RpcOutputStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            return StepDecision.gracefulComplete(2);
        }
    }
}
